File: sketch/src/main/java/me/xiaopan/sketch/cache/BitmapPoolUtils.java
/*
 * Copyright (C) 2016 Peng fei Pan <sky@xiaopan.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.sketch.cache;

import android.annotation.TargetApi;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Rect;
import android.os.Build;
import android.text.TextUtils;

import me.xiaopan.sketch.SLog;
import me.xiaopan.sketch.SLogType;
import me.xiaopan.sketch.SketchMonitor;
import me.xiaopan.sketch.decode.ImageType;
import me.xiaopan.sketch.util.SketchUtils;

public class BitmapPoolUtils {
    /**
     * SDK版本是否支持inBitmap，适用于BitmapFactory
     */
    public static boolean sdkSupportInBitmap() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB;
    }

    /**
     * SDK版本是否支持inBitmap，适用于BitmapRegionDecoder
     */
    public static boolean sdkSupportInBitmapForRegionDecoder() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN;
    }

    /**
     * 从bitmap poo中取出可复用的Bitmap设置到inBitmap上，适用于BitmapFactory
     *
     * @param options    BitmapFactory.Options 需要用到options的outWidth、outHeight、inSampleSize以及inPreferredConfig属性
     * @param bitmapPool BitmapPool 从这个池子里找可复用的Bitmap
     * @return true：找到了可复用的Bitmap
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public static boolean setInBitmapFromPool(BitmapFactory.Options options, BitmapPool bitmapPool) {
        if (!sdkSupportInBitmap()) {
            return false;
        }

        if (options.outWidth == 0 || options.outHeight == 0) {
            SLog.e(SLogType.REQUEST, "outWidth or ourHeight is 0");
            return false;
        }

        if (TextUtils.isEmpty(options.outMimeType)) {
            SLog.e(SLogType.REQUEST, "outMimeType is empty");
            return false;
        }

        // 使用inBitmap时4.4以下inSampleSize不能为0，最小也得是1
        if (options.inSampleSize <= 0) {
            options.inSampleSize = 1;
        }

        int inSampleSize = options.inSampleSize;
        ImageType imageType = ImageType.valueOfMimeType(options.outMimeType);

        Bitmap inBitmap = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            int finalWidth = SketchUtils.ceil(options.outWidth, inSampleSize);
            int finalHeight = SketchUtils.ceil(options.outHeight, inSampleSize);
            inBitmap = bitmapPool.get(finalWidth, finalHeight, options.inPreferredConfig);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB && inSampleSize == 1
                && (imageType == ImageType.JPEG || imageType == ImageType.PNG)) {
            inBitmap = bitmapPool.get(options.outWidth, options.outHeight, options.inPreferredConfig);
        }

        if (inBitmap != null && SLogType.CACHE.isEnabled()) {
            int sizeInBytes = SketchUtils.computeByteCount(options.outWidth, options.outHeight, options.inPreferredConfig);
            SLog.d(SLogType.CACHE, "setInBitmapFromPool. options=%dx%d,%s,%d,%d. inBitmap=%s,%d",
                    options.outWidth, options.outHeight, options.inPreferredConfig, inSampleSize, sizeInBytes,
                    Integer.toHexString(inBitmap.hashCode()), SketchUtils.getByteCount(inBitmap));
        }

        options.inBitmap = inBitmap;
        options.inMutable = true;

        return inBitmap != null;
    }

    public static boolean inBitmapThrow(Throwable throwable, BitmapFactory.Options options,
                                        SketchMonitor monitor, BitmapPool bitmapPool, String imageUri, int imageWidth, int imageHeight) {
        if (throwable instanceof IllegalArgumentException) {
            if (sdkSupportInBitmap()) {
                if (options.inBitmap != null) {
                    monitor.onInBitmapException(imageUri, imageWidth, imageHeight, options.inSampleSize, options.inBitmap);
                    freeBitmapToPool(options.inBitmap, bitmapPool);
                    return true;
                }
            }
        }

        return false;
    }

    /**
     * 处理bitmap，首先尝试放入bitmap pool，放不进去就回收
     *
     * @param bitmap     要处理的bitmap
     * @param bitmapPool BitmapPool 尝试放入这个池子
     * @return ture：成功放入bitmap pool
     */
    public static boolean freeBitmapToPool(Bitmap bitmap, BitmapPool bitmapPool) {
        if (bitmap == null || bitmap.isRecycled()) {
            return false;
        }

        boolean success = bitmapPool.put(bitmap);
        if (success) {
            if (SLogType.CACHE.isEnabled()) {
                StackTraceElement[] elements = new Exception().getStackTrace();
                StackTraceElement element = elements.length > 1 ? elements[1] : elements[0];
                SLog.d(SLogType.CACHE, String.format("Put to bitmap pool. info:%dx%d,%s,%s - %s.%s:%d",
                        bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig(), SketchUtils.toHexString(bitmap),
                        element.getClassName(), element.getMethodName(), element.getLineNumber()));
            }
        } else {
            if (SLogType.CACHE.isEnabled()) {
                StackTraceElement[] elements = new Exception().getStackTrace();
                StackTraceElement element = elements.length > 1 ? elements[1] : elements[0];
                SLog.w(SLogType.CACHE, String.format("Recycle bitmap. info:%dx%d,%s,%s - %s.%s:%d",
                        bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig(), SketchUtils.toHexString(bitmap),
                        element.getClassName(), element.getMethodName(), element.getLineNumber()));
            }
            bitmap.recycle();
        }
        return success;
    }

    /**
     * 从bitmap poo中取出可复用的Bitmap设置到inBitmap上，适用于BitmapRegionDecoder
     *
     * @param options    BitmapFactory.Options 需要用到options的outWidth、outHeight、inSampleSize以及inPreferredConfig属性
     * @param bitmapPool BitmapPool 从这个池子里找可复用的Bitmap
     * @return true：找到了可复用的Bitmap
     */
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public static boolean setInBitmapFromPoolForRegionDecoder(BitmapFactory.Options options, Rect srcRect, BitmapPool bitmapPool) {
        if (!sdkSupportInBitmapForRegionDecoder()) {
            return false;
        }

        int inSampleSize = options.inSampleSize >= 1 ? options.inSampleSize : 1;
        Bitmap.Config config = options.inPreferredConfig;

        int finalWidth = SketchUtils.ceil(srcRect.width(), inSampleSize);
        int finalHeight = SketchUtils.ceil(srcRect.height(), inSampleSize);
        Bitmap inBitmap = bitmapPool.get(finalWidth, finalHeight, config);

        if (inBitmap != null) {
            if (SLogType.CACHE.isEnabled()) {
                int sizeInBytes = SketchUtils.computeByteCount(finalWidth, finalHeight, config);
                SLog.d(SLogType.CACHE, "setInBitmapFromPoolForRegionDecoder. options=%dx%d,%s,%d,%d. inBitmap=%s,%d",
                        finalWidth, finalHeight, config, inSampleSize, sizeInBytes,
                        Integer.toHexString(inBitmap.hashCode()), SketchUtils.getByteCount(inBitmap));
            }
        } else {
            // 由于BitmapRegionDecoder不支持inMutable所以就自己创建Bitmap
            inBitmap = Bitmap.createBitmap(finalWidth, finalHeight, config);
        }

        options.inBitmap = inBitmap;

        return inBitmap != null;
    }

    public static boolean inBitmapThrowForRegionDecoder(Throwable throwable, BitmapFactory.Options options,
                                                        SketchMonitor monitor, BitmapPool bitmapPool, String imageUri,
                                                        int imageWidth, int imageHeight, Rect srcRect) {
        if (throwable instanceof IllegalArgumentException) {
            if (sdkSupportInBitmapForRegionDecoder()) {
                if (options.inBitmap != null) {
                    monitor.onInBitmapExceptionForRegionDecoder(imageUri, imageWidth,
                            imageHeight, srcRect, options.inSampleSize, options.inBitmap);
                    freeBitmapToPoolForRegionDecoder(options.inBitmap, bitmapPool);
                    return true;
                }
            }
        }

        return false;
    }

    /**
     * 处理bitmap，首先尝试放入bitmap pool，放不进去就回收
     *
     * @param bitmap     要处理的bitmap
     * @param bitmapPool BitmapPool 尝试放入这个池子
     * @return true：成功放入bitmap pool
     */
    public static boolean freeBitmapToPoolForRegionDecoder(Bitmap bitmap, BitmapPool bitmapPool) {
        if (bitmap == null || bitmap.isRecycled()) {
            return false;
        }

        boolean success = sdkSupportInBitmapForRegionDecoder() && bitmapPool.put(bitmap);
        if (!success) {
            if (SLogType.CACHE.isEnabled()) {
                StackTraceElement[] elements = new Exception().getStackTrace();
                StackTraceElement element = elements.length > 1 ? elements[1] : elements[0];
                SLog.w(SLogType.CACHE, String.format("Recycle bitmap. info:%dx%d,%s,%s - %s.%s:%d",
                        bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig(), SketchUtils.toHexString(bitmap),
                        element.getClassName(), element.getMethodName(), element.getLineNumber()));
            }
            bitmap.recycle();
        } else {
            if (SLogType.CACHE.isEnabled()) {
                StackTraceElement[] elements = new Exception().getStackTrace();
                StackTraceElement element = elements.length > 1 ? elements[1] : elements[0];
                SLog.d(SLogType.CACHE, String.format("Put to bitmap pool. info:%dx%d,%s,%s - %s.%s:%d",
                        bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig(), SketchUtils.toHexString(bitmap),
                        element.getClassName(), element.getMethodName(), element.getLineNumber()));
            }
        }
        return success;
    }
}

File: sketch/src/main/java/com/github/panpf/sketch/decode/internal/BitmapPoolUtils.kt
/*
 * Copyright (C) 2022 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.decode.internal

import android.graphics.Bitmap
import android.graphics.Bitmap.Config
import android.graphics.BitmapFactory.Options
import com.github.panpf.sketch.cache.BitmapPool
import com.github.panpf.sketch.util.Size
import com.github.panpf.sketch.util.isAndSupportHardware
import com.github.panpf.sketch.util.isMainThread
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

private const val MODULE = "BitmapPoolUtils"


fun BitmapPool.setInBitmap(
    options: Options,
    imageSize: Size,
    imageMimeType: String?,
    disallowReuseBitmap: Boolean = false,
    caller: String? = null,
): Boolean {
    if (disallowReuseBitmap) {
        logger?.w(MODULE) {
            "setInBitmap. disallowReuseBitmap. imageSize=$imageSize, imageMimeType=$imageMimeType. $caller"
        }
        return false
    }
    if (imageSize.isEmpty) {
        logger?.e(MODULE, "setInBitmap. error. imageSize is empty: $imageSize. $caller")
        return false
    }
    if (options.inPreferredConfig?.isAndSupportHardware() == true) {
        logger?.e(MODULE) {
            "setInBitmap. error. inPreferredConfig is HARDWARE does not support inBitmap. $caller"
        }
        return false
    }

    options.inMutable = true

    val inSampleSize = options.inSampleSize.coerceAtLeast(1)
    if (!isSupportInBitmap(imageMimeType, inSampleSize)) {
        logger?.w(MODULE) {
            "setInBitmap. error. " +
                    "The current configuration does not support the use of inBitmap in BitmapFactory. " +
                    "imageMimeType=$imageMimeType, inSampleSize=${options.inSampleSize}. " +
                    "For details, please refer to 'DecodeUtils.isSupportInBitmap()'. " +
                    "$caller"
        }
        return false
    }
    val sampledBitmapSize =
        calculateSampledBitmapSize(imageSize, inSampleSize, imageMimeType)
    val inBitmap: Bitmap? = get(
        width = sampledBitmapSize.width,
        height = sampledBitmapSize.height,
        config = options.inPreferredConfig
    )
    if (inBitmap != null) {
        logger?.d(MODULE) {
            "setInBitmap. successful. " +
                    "imageSize=$imageSize, inSampleSize=$inSampleSize, imageMimeType=$imageMimeType. " +
                    "inBitmap=${inBitmap.logString}. " +
                    "$caller"
        }
    } else {
        logger?.d(MODULE) {
            "setInBitmap. failed. " +
                    "imageSize=$imageSize, inSampleSize=$inSampleSize, imageMimeType=$imageMimeType. " +
                    "$caller"
        }
    }

    // IllegalArgumentException("Problem decoding into existing bitmap") is thrown when inSampleSize is 0 but inBitmap is not null
    options.inSampleSize = inSampleSize
    options.inBitmap = inBitmap
    return inBitmap != null
}

fun BitmapPool.setInBitmapForRegion(
    options: Options,
    regionSize: Size,
    imageMimeType: String?,
    imageSize: Size,
    disallowReuseBitmap: Boolean = false,
    caller: String? = null,
): Boolean {
    if (disallowReuseBitmap) {
        logger?.w(MODULE) {
            "setInBitmapForRegion. disallowReuseBitmap. imageSize=$imageSize, imageMimeType=$imageMimeType. $caller"
        }
        return false
    }
    if (regionSize.isEmpty) {
        logger?.e(
            MODULE,
            "setInBitmapForRegion. error. regionSize is empty: $regionSize. $caller"
        )
        return false
    }
    if (options.inPreferredConfig?.isAndSupportHardware() == true) {
        logger?.e(MODULE) {
            "setInBitmapForRegion. error. inPreferredConfig is HARDWARE does not support inBitmap. $caller"
        }
        return false
    }
    if (!isSupportInBitmapForRegion(imageMimeType)) {
        logger?.w(MODULE) {
            "setInBitmapForRegion. error. " +
                    "The current configuration does not support the use of inBitmap in BitmapFactory. " +
                    "imageMimeType=$imageMimeType. " +
                    "For details, please refer to 'DecodeUtils.isSupportInBitmapForRegion()'. " +
                    "$caller"
        }
        return false
    }

    val inSampleSize = options.inSampleSize.coerceAtLeast(1)
    val sampledBitmapSize = calculateSampledBitmapSizeForRegion(
        regionSize, inSampleSize, imageMimeType, imageSize
    )
    // BitmapRegionDecoder does not support inMutable, so creates Bitmap
    var newCreate = false
    val inBitmap = get(
        sampledBitmapSize.width, sampledBitmapSize.height, options.inPreferredConfig
    ) ?: Bitmap.createBitmap(
        sampledBitmapSize.width, sampledBitmapSize.height, options.inPreferredConfig
    )!!.apply {
        newCreate = true
    }
    logger?.d(MODULE) {
        "setInBitmapForRegion. successful. " +
                "newCreate $newCreate. " +
                "regionSize=$regionSize, inSampleSize=$inSampleSize, imageSize=$imageSize. " +
                "inBitmap=${inBitmap.logString}. " +
                "$caller"
    }

    // IllegalArgumentException("Problem decoding into existing bitmap") is thrown when inSampleSize is 0 but inBitmap is not null
    options.inSampleSize = inSampleSize
    options.inBitmap = inBitmap
    return true
}

fun BitmapPool.getOrCreate(
    width: Int,
    height: Int,
    config: Config,
    disallowReuseBitmap: Boolean = false,
    caller: String? = null,
): Bitmap {
    if (disallowReuseBitmap) {
        return Bitmap.createBitmap(width, height, config).apply {
            logger?.d(MODULE) {
                "getOrCreate. new disallowReuseBitmap. ${this.logString}. $caller"
            }
        }
    }
    return get(width, height, config)
        ?: Bitmap.createBitmap(width, height, config)
            .apply {
                logger?.d(MODULE) {
                    "getOrCreate. new . ${this.logString}. $caller"
                }
            }
}

@OptIn(DelicateCoroutinesApi::class)
fun BitmapPool.freeBitmap(
    bitmap: Bitmap?,
    disallowReuseBitmap: Boolean = false,
    caller: String? = null,
) {
    if (bitmap != null && !bitmap.isRecycled) {
        if (!disallowReuseBitmap) {
            if (isMainThread()) {
                GlobalScope.launch(Dispatchers.IO) {
                    @Suppress("KotlinConstantConditions")
                    freeBitmap(bitmap, disallowReuseBitmap, caller)
                }
            } else {
                val success = put(bitmap, caller)
                if (success) {
                    logger?.d(MODULE) {
                        "freeBitmap. successful. $caller. ${bitmap.logString}"
                    }
                } else {
                    bitmap.recycle()
                    logger?.w(MODULE) {
                        "freeBitmap. failed. execute recycle. $caller. ${bitmap.logString}"
                    }
                }
            }
        } else {
            bitmap.recycle()
            logger?.w(MODULE) {
                "freeBitmap. disallowReuseBitmap. execute recycle. $caller. ${bitmap.logString}"
            }
        }
    } else {
        logger?.w(MODULE) {
            "freeBitmap. error. bitmap null or recycled. $caller. ${bitmap?.logString}"
        }
    }
}